internal/test-framework/pkg/runner/container/container.go

```diff
@@ -0,0 +1,285 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// To disable the `stutter` check
+// nolint: golint
+package container
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/talos-systems/talos/internal/test-framework/pkg/checker"
+	"github.com/talos-systems/talos/internal/test-framework/pkg/runner"
+	"github.com/talos-systems/talos/pkg/retry"
+)
+
+// ContainerRunner implements the Runner interface.
+// ContainerRunner contains the container configuration and client.
+type ContainerRunner struct {
+	ClusterName string
+	Container   *container.Config
+	Host        *container.HostConfig
+	Client      *client.Client
+}
+
+// New creates a new ContainerRunner with the specified options and
+// pulls the necessary image.
+func New(setters ...Option) (runner.Runner, error) {
+	var (
+		err    error
+		reader io.ReadCloser
+		result *multierror.Error
+	)
+
+	cr := defaultOptions()
+
+	for _, setter := range setters {
+		result = multierror.Append(result, setter(cr))
+	}
+
+	if reader, err = cr.Client.ImagePull(context.Background(), cr.Container.Image, types.ImagePullOptions{}); err != nil {
+		result = multierror.Append(result, err)
```

Comment:
I think there's a flow problem here:

1. if method returns an error, `reader` is undefined, so we should abort execution here, we can't really continue
2. in fact, same above for setting options: if options can't be set, we shouldn't keep running 

so I believe instead of `multierror` we should immediately fail on error